##Step1: Import Libraries

In [2]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

from sklearn.metrics import classification_report, confusion_matrix, roc_auc_score

import joblib
print("Libraries imported successfully!")

Libraries imported successfully!


##Step2:Load Datasets

In [4]:
from google.colab import files
uploaded = files.upload()

Saving archive (2).zip to archive (2).zip


In [8]:
import os
os.listdir()

['.config', 'archive (2).zip', 'sample_data']

##Step3: Unzip the file

In [10]:
import zipfile

with zipfile.ZipFile("archive (2).zip", 'r') as zip_ref:
    zip_ref.extractall()
    print("Dataset extracted successfully!")

Dataset extracted successfully!


##Step4: Verify Extracted file

In [11]:
import os
os.listdir()

['.config',
 'fraudTest.csv',
 'fraudTrain.csv',
 'data',
 'archive (2).zip',
 'sample_data']

##Step5: Data Loading

In [12]:
import pandas as pd

df = pd.read_csv("fraudTrain.csv")
df.head()

,Unnamed: 0,trans_date_trans_time,cc_num,merchant,category,amt,first,last,gender,street,...,lat,long,city_pop,job,dob,trans_num,unix_time,merch_lat,merch_long,is_fraud
0,0,2019-01-01 00:00:18,2703186189652095,"fraud_Rippin, Kub and Mann",misc_net,4.97,Jennifer,Banks,F,561 Perry Cove,...,36.0788,-81.1781,3495,"Psychologist, counselling",1988-03-09,0b242abb623afc578575680df30655b9,1325376018,36.011293,-82.048315,0
1,1,2019-01-01 00:00:44,630423337322,"fraud_Heller, Gutmann and Zieme",grocery_pos,107.23,Stephanie,Gill,F,43039 Riley Greens Suite 393,...,48.8878,-118.2105,149,Special educational needs teacher,1978-06-21,1f76529f8574734946361c461b024d99,1325376044,49.159047,-118.186462,0
2,2,2019-01-01 00:00:51,38859492057661,fraud_Lind-Buckridge,entertainment,220.11,Edward,Sanchez,M,594 White Dale Suite 530,...,42.1808,-112.2620,4154,Nature conservation officer,1962-01-19,a1a22d70485983eac12b5b88dad1cf95,1325376051,43.150704,-112.154481,0
3,3,2019-01-01 00:01:16,3534093764340240,"fraud_Kutch, Hermiston and Farrell",gas_transport,45.00,Jeremy,White,M,9443 Cynthia Court Apt. 038,...,46.2306,-112.1138,1939,Patent attorney,1967-01-12,6b849c168bdad6f867558c3793159a81,1325376076,47.034331,-112.561071,0
4,4,2019-01-01 00:03:06,375534208663984,fraud_Keeling-Crist,misc_pos,41.96,Tyler,Garcia,M,408 Bradley Rest,...,38.4207,-79.4629,99,Dance movement psychotherapist,1986-03-28,a41d7549acf90789359a9aa5346dcb46,1325376186,38.674999,-78.632459,0


##Step6: Exploratory Data Analysis (EDA)

In [14]:
df = pd.read_csv("fraudTrain.csv")

print("Rows and Columns:")
print(df.shape)

print("\nFirst 5 Rows:")
print(df.head())

print("\nMissing Values:")
print(df.isnull().sum())

print("\nFraud vs Legit Transactions:")
print(df["is_fraud"].value_counts())

Rows and Columns:
(1296675, 23)

First 5 Rows:
   Unnamed: 0 trans_date_trans_time            cc_num  \
0           0   2019-01-01 00:00:18  2703186189652095   
1           1   2019-01-01 00:00:44      630423337322   
2           2   2019-01-01 00:00:51    38859492057661   
3           3   2019-01-01 00:01:16  3534093764340240   
4           4   2019-01-01 00:03:06   375534208663984   

                             merchant       category     amt      first  \
0          fraud_Rippin, Kub and Mann       misc_net    4.97   Jennifer   
1     fraud_Heller, Gutmann and Zieme    grocery_pos  107.23  Stephanie   
2                fraud_Lind-Buckridge  entertainment  220.11     Edward   
3  fraud_Kutch, Hermiston and Farrell  gas_transport   45.00     Jeremy   
4                 fraud_Keeling-Crist       misc_pos   41.96      Tyler   

      last gender                        street  ...      lat      long  \
0    Banks      F                561 Perry Cove  ...  36.0788  -81.1781   
1     Gil

##Step7: Data Cleaning

In [22]:
print(df.columns)

Index(['cc_num', 'merchant', 'category', 'amt', 'gender', 'state', 'lat',
       'long', 'city_pop', 'job', 'unix_time', 'merch_lat', 'merch_long',
       'is_fraud'],
      dtype='object')


In [23]:
df = df.drop(['Unnamed: 0', 'trans_date_trans_time', 'first', 'last',
               'street', 'city', 'zip', 'dob', 'trans_num'],
              axis=1, errors='ignore')

##Step8: Label Encoding (Categorical Data Conversion)

In [27]:
le = LabelEncoder()

df['merchant'] = le.fit_transform(df['merchant'])
df['category'] = le.fit_transform(df['category'])
df['gender'] = le.fit_transform(df['gender'])
df['state'] = le.fit_transform(df['state'])
df['job'] = le.fit_transform(df['job'])
print(df.head())

             cc_num  merchant  category     amt  gender  state      lat  \
0  2703186189652095       514         8    4.97       0     27  36.0788   
1      630423337322       241         4  107.23       0     47  48.8878   
2    38859492057661       390         0  220.11       1     13  42.1808   
3  3534093764340240       360         2   45.00       1     26  46.2306   
4   375534208663984       297         9   41.96       1     45  38.4207   

       long  city_pop  job   unix_time  merch_lat  merch_long  is_fraud  
0  -81.1781      3495  370  1325376018  36.011293  -82.048315         0  
1 -118.2105       149  428  1325376044  49.159047 -118.186462         0  
2 -112.2620      4154  307  1325376051  43.150704 -112.154481         0  
3 -112.1138      1939  328  1325376076  47.034331 -112.561071         0  
4  -79.4629        99  116  1325376186  38.674999  -78.632459         0  


##Step9: Feature Selection (X) and Target Separation (y)

In [28]:
X = df.drop('is_fraud', axis=1)
y = df['is_fraud']

print("Features shape:", X.shape)
print("Target shape:", y.shape)

Features shape: (1296675, 13)
Target shape: (1296675,)


##Step10: Train-Test Split

In [29]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

print("Training data shape:", X_train.shape)
print("Testing data shape:", X_test.shape)

Training data shape: (1037340, 13)
Testing data shape: (259335, 13)


##Step11: Model Training (Machine Learning Models)
###Logistic Regression

In [33]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report

lr = LogisticRegression(max_iter=1000, class_weight='balanced')

lr.fit(X_train, y_train)

lr_pred = lr.predict(X_test)

print("Logistic Regression Results (Balanced):")
print(classification_report(y_test, lr_pred))

Logistic Regression Results (Balanced):
              precision    recall  f1-score   support

           0       0.99      1.00      1.00    257834
           1       0.00      0.00      0.00      1501

    accuracy                           0.99    259335
   macro avg       0.50      0.50      0.50    259335
weighted avg       0.99      0.99      0.99    259335



/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


###Decision Tree

In [34]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report, confusion_matrix

dt = DecisionTreeClassifier(random_state=42, class_weight='balanced')

dt.fit(X_train, y_train)

dt_pred = dt.predict(X_test)

print("Decision Tree Results:")
print(classification_report(y_test, dt_pred))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, dt_pred))

Decision Tree Results:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00    257834
           1       0.67      0.64      0.65      1501

    accuracy                           1.00    259335
   macro avg       0.83      0.82      0.82    259335
weighted avg       1.00      1.00      1.00    259335


Confusion Matrix:
[[257358    476]
 [   546    955]]


###Random Forest

In [39]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, confusion_matrix

rf = RandomForestClassifier(
    n_estimators=50,
    random_state=42,
    class_weight='balanced'
)

rf.fit(X_train, y_train)

rf_pred = rf.predict(X_test)

print("Random Forest Results:")
print(classification_report(y_test, rf_pred))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, rf_pred))

Random Forest Results:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00    257834
           1       0.89      0.64      0.74      1501

    accuracy                           1.00    259335
   macro avg       0.95      0.82      0.87    259335
weighted avg       1.00      1.00      1.00    259335


Confusion Matrix:
[[257722    112]
 [   547    954]]


##Step12: Evaluate Model

In [40]:
from sklearn.metrics import roc_auc_score

print("ROC-AUC Scores:")

print("Logistic Regression:", roc_auc_score(y_test, lr_pred))
print("Decision Tree:", roc_auc_score(y_test, dt_pred))
print("Random Forest:", roc_auc_score(y_test, rf_pred))

ROC-AUC Scores:
Logistic Regression: 0.5
Decision Tree: 0.8171981779620049
Random Forest: 0.8175709472306257


##Step13: Compare Models

In [41]:
scores = {
    "Logistic Regression": roc_auc_score(y_test, lr_pred),
    "Decision Tree": roc_auc_score(y_test, dt_pred),
    "Random Forest": roc_auc_score(y_test, rf_pred)
}

best_model = max(scores, key=scores.get)

print("Best Model is:", best_model)

Best Model is: Random Forest


##Step14:Choose the Best Model

In [43]:
lr_score = float(roc_auc_score(y_test, lr_pred))
dt_score = float(roc_auc_score(y_test, dt_pred))
rf_score = float(roc_auc_score(y_test, rf_pred))

scores = {
    "Logistic Regression": lr_score,
    "Decision Tree": dt_score,
    "Random Forest": rf_score
}

best_model = max(scores, key=scores.get)

print("Scores:", scores)
print("Best Model is:", best_model)

Scores: {'Logistic Regression': 0.5, 'Decision Tree': 0.8171981779620049, 'Random Forest': 0.8175709472306257}
Best Model is: Random Forest


##Step15:Save the model

In [45]:
import joblib

joblib.dump(rf, "fraud_detection_model.pkl")

print("Model saved successfully!")

Model saved successfully!


##Step 16: Download

In [71]:
from google.colab import files
files.download("fraud_detection_model.pkl")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

##Step17: Tests

In [48]:
import joblib

model = joblib.load("fraud_detection_model.pkl")
print("Model loaded successfully!")

Model loaded successfully!


###Test_1

In [70]:
sample1 = X_test[y_test == 1].iloc[0:1]
prediction1 = rf.predict(sample1)

if prediction1[0] == 1:
    print("Test 1: FRAUD")
else:
    print("Test 1: LEGIT")

Test 1: LEGIT


###Test_2

In [67]:
sample2 = X_test[y_test == 1].iloc[1:2]
prediction = rf.predict(sample2)

if prediction[0] == 1:
    print("Test 2: FRAUD")
else:
    print("Test 2: LEGIT")

Test 2: FRAUD


###Test_3

In [68]:
sample3 = X_test[y_test == 1].iloc[2:3]
prediction = rf.predict(sample3)

if prediction[0] == 1:
    print("Test 3: FRAUD")
else:
    print("Test 3: MISSED FRAUD")

Test 3: FRAUD


###Conclusion:

####In this task, I built a Credit Card Fraud Detection system using Machine Learning. The dataset was preprocessed by handling missing values, encoding categorical variables, and splitting the data into training and testing sets.

####Three machine learning models were trained and compared:

####1. Logistic Regression
####2. Decision Tree
####3. Random Forest

###Result:
####The Random Forest model was selected as the final model because it gave the best overall performance and higher accuracy 81.76%  in identifying fraudulent transactions.

####It was able to successfully distinguish between fraudulent and legitimate transactions, making it suitable for real-world fraud detection systems.